measurements/api/measurements.py

```diff
@@ -420,171 +390,226 @@ def list_measurements(
+
     if anomaly is not None:
-        fpq = fpq.filter(
-            cast(cast(Fastpath.scores["blocking_general"], String), Float) > 0.5
+        # TODO: coalesce false() ?
+        query_params["anomaly"] = anomaly
+        mrwhere.append(sql.text("measurement.anomaly = :anomaly"))
+        # fpwhere.append(sql.text("scores->>'blocking_general' > 0.5"))
+        fpwhere.append(sql.text("false"))
+
+    fpq_table = sql.table("fastpath")
+    mr_table = sql.table("measurement").join(
+        sql.table("report"), sql.text("measurement.report_no = report.report_no"),
+    )
+
+    if input_ or domain or category_code:
+        # join in domain_input
+        mr_table = mr_table.join(
+            sql.table("domain_input"),
+            sql.text("domain_input.input_no = measurement.input_no"),
+        )
+        fpq_table = fpq_table.join(
+            sql.table("domain_input"), sql.text("domain_input.input = fastpath.input")
         )
-    if order_by is not None:
-        q = q.order_by(text("{} {}".format(order_by, order)))
-    q = q.limit(limit).offset(offset)
-    # Assemble the query
 
-    try:
-        if order_by is not None:
-            fpq = fpq.order_by(text("{} {}".format(order_by, order)))
-        fpq = fpq.limit(limit).offset(offset)
-        q = q.union(fpq)
-    except:
-        log_query(log, q)
-        log_query(log, fpq)
-        raise
+        if input_:
+            # input_ overrides domain and category_code
+            query_params["input"] = input_
+            mrwhere.append(sql.text("domain_input.input = :input"))
+            fpwhere.append(sql.text("domain_input.input = :input"))
+
+        else:
+            # both domain and category_code can be set at the same time
+            if domain:
+                query_params["domain"] = domain
+                mrwhere.append(sql.text("domain_input.domain = :domain"))
+                fpwhere.append(sql.text("domain_input.domain = :domain"))
+
+            if category_code:
+                query_params["category_code"] = category_code
+                mr_table = mr_table.join(
+                    sql.table("citizenlab"),
+                    sql.text("citizenlab.url = domain_input.input"),
+                )
+                fpq_table = fpq_table.join(
+                    sql.table("citizenlab"),
+                    sql.text("citizenlab.url = domain_input.input"),
+                )
+                mrwhere.append(sql.text("citizenlab.category_code = :category_code"))
+                fpwhere.append(sql.text("citizenlab.category_code = :category_code"))
 
+    else:
+        mr_table = mr_table.join(
+            sql.table("domain_input"),
+            sql.text("domain_input.input_no = measurement.input_no"),
+        )
+
+    # We runs SELECTs on the measurement-report (mr) tables and faspath independently
+    # from each other and then merge them.
+    # An ORDER BY + LIMIT on "limit+offset" is applied in each SELECT as a speedup.
```

Comment:
I wonder if an additional few words are necessary to clarify the significance of using offset + limit, as it was not obvious to me in the beginning we should do that.

Something to say that we need to retrieve all the records up to limit + offset and then apply the offset and limit on the outer query.

Reply:
Added comment in ff7529b188a622bdcde80aed2f56ec7f39d9dbc0